### PR TITLE
Grpc read optimization to not prematurely close existing requests

### PR DESCRIPTION
**Pull request by abmodi**
Hadoop FsDataInputStream exposes read APIs which will call seek back to original position after the read. Generally they are followed by another seek before actual read happens. We should only close the requests when we are actually reading rather than on seeks.

### CODE REVIEW

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcReadChannel.java [hunk 1]
```diff
@@ -389,6 +382,22 @@ private int readBufferedContentInto(ByteBuffer byteBuffer) {
     return bytesToWrite;
   }
 
+  private boolean canReadFromExistingRequest(ByteBuffer byteBuffer) {
```
Comment (arunkumarchacko) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcReadChannel.java:385]: can we have the private methods after the public methods? Will help with readability.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcReadChannel.java [hunk 2]
```diff
@@ -389,6 +382,22 @@ private int readBufferedContentInto(ByteBuffer byteBuffer) {
...
+    if (resIterator == null) {
+      return false;
+    }
+
+    if (positionForNextRead < positionInGrpcStream) {
```
Comment (arunkumarchacko) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcReadChannel.java:390]: when can this happen?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcReadChannel.java [hunk 3]
```diff
@@ -389,6 +382,22 @@ private int readBufferedContentInto(ByteBuffer byteBuffer) {
     return bytesToWrite;
   }
 
+  private boolean canReadFromExistingRequest(ByteBuffer byteBuffer) {
+    if (resIterator == null) {
```
Comment (arunkumarchacko) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcReadChannel.java:386]: does this method need to be synchronised? i.e. can resIterator be set to null from another thread? Same for other instance fields used.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcReadChannel.java [hunk 4]
```diff
@@ -725,30 +728,17 @@ public SeekableByteChannel position(long newPosition) throws IOException {
...
     if (readStrategy == Fadvise.AUTO) {
-      if (seekDistance < 0 || seekDistance > readOptions.getInplaceSeekLimit()) {
+      if (positionForNextRead < positionInGrpcStream
+          || positionForNextRead - positionInGrpcStream > readOptions.getInplaceSeekLimit()) {
         readStrategy = Fadvise.RANDOM;
       }
     }
-
-    // Reset any ongoing read operations or local data caches.
-    cancelCurrentRequest();
```
Comment (arunkumarchacko) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcReadChannel.java:747]: Is it ok to remove this? i.e. can b/210660938 regress?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcReadChannel.java (hunk 2 above)
Comment (abmodi) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcReadChannel.java:390]: In case of backward seek.

**Review by mprashanthsagar — APPROVED** [review]

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcReadChannel.java [hunk 5]
```diff
@@ -363,11 +357,10 @@ private static void put(ByteString source, int offset, int size, ByteBuffer dest
 
   private int readBufferedContentInto(ByteBuffer byteBuffer) {
     // Handle skipping forward through the buffer for a seek.
-    long bufferSkip =
-        min(bufferedContent.size() - bufferedContentReadOffset, bytesToSkipBeforeReading);
+    long bytesToSkip = positionForNextRead - positionInGrpcStream;
+    long bufferSkip = min(bufferedContent.size() - bufferedContentReadOffset, bytesToSkip);
```
Comment (mayanks) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcReadChannel.java:361]: while we are at it (simplifying the logic) we can look at getting rid of `bufferedContentReadOffset` variable as well. We can ensure that `bufferedContent` always has relevant data from offset 0. Else we can  do 
```
   bufferedContent = bufferedContent.substring(skipBytes)
```
This reduces the complexity of maintaining both `bufferedContent` and `bytesToSkipBeforeReading`

**Review by mayanks — APPROVED** [review]

**Pull request merged**